sdk/python/lib/pulumi/runtime/resource.py

```diff
@@ -138,8 +139,62 @@ async def prepare_resource(res: 'Resource',
     )
 
 
-def read_resource(res: 'CustomResource', ty: str, name: str, props: 'Inputs', opts: 'ResourceOptions') -> None:
+def resource_output(res: 'Resource') -> Tuple[Callable[[Any, bool, bool, Optional[Exception]], None], 'Output']:
     from .. import Output  # pylint: disable=import-outside-toplevel
+
+    value_future: asyncio.Future[Any] = asyncio.Future()
+    known_future: asyncio.Future[bool] = asyncio.Future()
+    secret_future: asyncio.Future[bool] = asyncio.Future()
+
+    def resolve(value: Any, known: bool, secret: bool, exn: Optional[Exception]):
+        if exn is not None:
+            value_future.set_exception(exn)
+            known_future.set_exception(exn)
+            secret_future.set_exception(exn)
+        else:
+            value_future.set_result(value)
+            known_future.set_result(known)
+            secret_future.set_result(secret)
+
+    return (resolve, Output({res}, value_future, known_future, secret_future))
+
+
+def resolved_resource_output(res: 'Resource', value: Any, known: bool, secret: bool) -> 'Output':
+    (resolve, output) = resource_output(res)
+    resolve(value, known, secret, None)
+    return output
+
+
+def get_resource(res: 'Resource', custom: bool, urn: str) -> None:
+    log.debug(f"getting resource: urn={urn}")
+
+    result = invoke("pulumi:pulumi:getResource", {"urn": urn})
+
+    # Same as below, we initialize the URN property on the resource, which will always be resolved.
+    log.debug("preparing get resource for RPC")
+
+    res.__dict__["urn"] = resolved_resource_output(res, urn, True, False)
+
+    # If this is a custom resource, resolve the ID.
+    if custom:
+        id = result.value["id"]
+        id_known = bool(id)
+
+        res.__dict__["id"] = resolved_resource_output(res, id, id_known, False)
+
+    # Create resolved outputs for each property on res.
+    for key, value in result.value["state"].items():
+        # Skip well-known properties.
+        if key == "urn" or key == "id":
+            continue
+
+        is_known = not rpc.contains_unknowns(value)
+        is_secret = rpc.is_rpc_secret(value)
+
+        res.__dict__[key] = resolved_resource_output(res, rpc.unwrap_rpc_secret(value), is_known, is_secret)
```

Comment:
Don't we need to translate the key into the Python name and also specify the intended type? I think we need to do something like:

https://github.com/pulumi/pulumi/blob/f22b47e54f39312b9b13ea359d3d523ee42e7a82/sdk/python/lib/pulumi/runtime/rpc.py#L571-L578

Reply:
I think that's already been done by the deserializer, hasn't it?

Reply:
No, I don't think so. When I implemented support for dataclasses, it was much simpler to implement the type conversion inside `translate_output_properties` rather than inside `deserialize_properties` (where you'd expect that code to live). So unfortunately, `translate_output_properties` is currently doing two things: 1. Translating key names (as it did originally), and 2. Translating the value to the target type, if specified. (Kinda gross, I know).

You can see inside the `invoke` call, all it's doing this:

https://github.com/pulumi/pulumi/blob/f22b47e54f39312b9b13ea359d3d523ee42e7a82/sdk/python/lib/pulumi/runtime/invoke.py#L125-L130

The `typ` parameter is an optional parameter to `invoke`. Our codegen for functions specifies `typ` when calling `invoke`, which takes care of translating the keys and instantiating the result class.

For example, in `pulumi_aws`, `get_bucket` does [this](https://github.com/pulumi/pulumi-aws/blob/2825d208fcf76bc890c5c2f5bb45921f896dc338/sdk/python/pulumi_aws/s3/get_bucket.py#L186):

```python
__ret__ = pulumi.runtime.invoke('aws:s3/getBucket:getBucket', __args__, opts=opts, typ=GetBucketResult).
```

`GetBucketResult` is decorated with `@pulumi.output_type`, so `translate_output_properties` will end up instantiating it and translating the keys based on the decorators on the class's properties.

In this particular case, for `getResource`, passing a `typ` to this invoke isn't going to work, because the target type is a resource, not an output type. So instead, I think we're going to need to do something along the lines of that code snippet in my earlier comment.

Basically call `types = _types.resource_types(type(res)) ` to get a dict of types for the resource, and then loop over the items in `state` translating the key with `res.translate_output_property(key)` and value with `translate_output_properties(value, res.translate_output_property, types.get(key)) `.

Reply:
Ah, no, you're right--the `invoke` deserializer won't do the translation. I may need to resort to a lower-level invoke s.t. I can call the translator.